### PR TITLE
fix: misleading error message and improves check_remote safeguard.

### PR DESCRIPTION
**Pull request by i4ki**
Now it reports the following error message:

```
$ terramate run ls
2022-11-03T13:04:07Z FTL current HEAD is out-of-date with the remote base branch: Please merge the latest changes from origin/main into this branch
```

### CODE REVIEW

File: cmd/terramate/cli/cli.go [hunk 1]
```diff
@@ -812,10 +812,6 @@ func (c *cli) format() {
 }
 
 func (c *cli) printStacks() {
-	if c.parsedArgs.Changed {
```
Comment (i4ki) [inline cmd/terramate/cli/cli.go:815]: @mariux I'm not sure if you want this removed now or later. Please let me know.

File: cmd/terramate/cli/cli.go [hunk 2]
```diff
@@ -676,6 +660,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
...
+		Str("action", "checkGit()").
+		Logger()
+
+	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
+		return
+	}
+
+	logger.Trace().Msg("check git default branch is updated")
+
+	if err := c.prj.checkLocalDefaultIsUpdated(); err != nil {
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:674]: the naming of the functions is very confusing.. what does this check do?

File: cmd/terramate/cli/cli.go [hunk 3]
```diff
@@ -676,6 +660,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
 	}
 }
 
+func (c *cli) gitRemoteSafeguard() {
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:663]:
```suggestion
func (c *cli) gitSafeguardDefaultBranchIsReachable() {
```

**Review by mariux — CHANGES_REQUESTED** [review]

File: cmd/terramate/cli/cli.go (hunk 2 above)
Comment (i4ki) [reply cmd/terramate/cli/cli.go]: I improved the name.

File: cmd/terramate/e2etests/safeguard_test.go [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+	git.CommitAll("add tempfile3")
+
+	testrun()
+
+	git.Checkout("HEAD^1")
+
+	testrun()
+}
+
+func TestSafeguardDisableGitCheckRemoteFromArgs(t *testing.T) {
```
Comment (katcipis) [inline cmd/terramate/e2etests/safeguard_test.go:118]:
```suggestion
func TestSafeguardDisableGitCheckRemote(t *testing.T) {
```

We have subtests to disable in different ways, not just args.

**Review by katcipis — APPROVED** [review]

**Review by katcipis — DISMISSED** [review]

File: cmd/terramate/cli/cli.go [hunk 5]
```diff
@@ -676,6 +660,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
...
+func (c *cli) gitSafeguardDefaultBranchIsReachable() {
+	logger := log.With().
+		Str("action", "checkGit()").
+		Logger()
+
+	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
+		return
+	}
+
+	logger.Trace().Msg("check git default branch is reachable")
```
Comment (i4ki) [inline cmd/terramate/cli/cli.go:672]:
```suggestion
	logger.Trace().Msg("check git default remote branch is reachable")
```

File: cmd/terramate/cli/cli.go (hunk 1 above)
Comment (mariux) [reply cmd/terramate/cli/cli.go:815]: yes, can be dropped.

**Review by katcipis — DISMISSED** [review]

File: cmd/terramate/cli/project.go [hunk 6]
```diff
@@ -225,24 +226,37 @@ func (p project) checkDefaultRemote() error {
...
+
+	remoteDesc := fmt.Sprintf("remote(%s/%s)", gitcfg.DefaultRemote, gitcfg.DefaultBranch)
+
+	logger := log.With().
+		Str("HEAD", p.headCommit()).
+		Str(remoteDesc, p.remoteDefaultCommit()).
+		Logger()
+
+	handleError := func(logger zerolog.Logger, debugErrCtx error) error {
+		logger.Debug().Msgf("%s: %v", ErrCurrentHeadIsOutOfDate, debugErrCtx)
```
Comment (mariux) [inline cmd/terramate/cli/project.go:240]: no need to duplicate same message in different levels..

```suggestion
		logger.Debug().Msgf("%v", debugErrCtx)
```

File: cmd/terramate/cli/project.go [hunk 7]
```diff
@@ -225,24 +226,37 @@ func (p project) checkDefaultRemote() error {
...
 
-func (p *project) checkLocalDefaultIsUpdated() error {
+func (p *project) checkRemoteDefaultBranchIsReachable() error {
 	gitcfg := p.gitcfg()
+
+	remoteDesc := fmt.Sprintf("remote(%s/%s)", gitcfg.DefaultRemote, gitcfg.DefaultBranch)
+
+	logger := log.With().
+		Str("HEAD", p.headCommit()).
+		Str(remoteDesc, p.remoteDefaultCommit()).
```
Comment (mariux) [inline cmd/terramate/cli/project.go:236]: context fields should not have special chars.

```suggestion
		Str("default_branch", remoteDesc).
		Str("default_hash", p.remoteDefaultCommit()).
```

File: cmd/terramate/cli/project.go [hunk 8]
```diff
@@ -225,24 +226,37 @@ func (p project) checkDefaultRemote() error {
...
+		Logger()
+
+	handleError := func(logger zerolog.Logger, debugErrCtx error) error {
+		logger.Debug().Msgf("%s: %v", ErrCurrentHeadIsOutOfDate, debugErrCtx)
+		return errors.E(
+			ErrCurrentHeadIsOutOfDate,
+			"Please merge the latest changes from %s into this branch",
```
Comment (mariux) [inline cmd/terramate/cli/project.go:243]: why is this not one error message?

**Review by mariux — CHANGES_REQUESTED** [review]

File: cmd/terramate/cli/project.go [hunk 9]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
+	remoteDesc := fmt.Sprintf("remote(%s/%s)", gitcfg.DefaultRemote, gitcfg.DefaultBranch)
+
+	logger := log.With().
+		Str("HEAD", p.headCommit()).
+		Str("default_branch", remoteDesc).
+		Str("default_hash", p.remoteDefaultCommit()).
+		Logger()
+
+	outOfDateErr := errors.E(
+		ErrCurrentHeadIsOutOfDate,
```
Comment (mariux) [inline cmd/terramate/cli/project.go:240]: why do we have error messages somewhere else? what is the benefit?

File: cmd/terramate/cli/project.go [hunk 10]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
 }
 
-func (p *project) checkLocalDefaultIsUpdated() error {
+func (p *project) checkRemoteDefaultBranchIsReachable() error {
 	gitcfg := p.gitcfg()
+
+	remoteDesc := fmt.Sprintf("remote(%s/%s)", gitcfg.DefaultRemote, gitcfg.DefaultBranch)
+
+	logger := log.With().
+		Str("HEAD", p.headCommit()).
```
Comment (mariux) [inline cmd/terramate/cli/project.go:234]:
```suggestion
		Str("head_hash", p.headCommit()).
```

File: cmd/terramate/cli/project.go [hunk 11]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
+
 	mergeBaseCommitID, err := p.git.wrapper.MergeBase(p.headCommit(), p.remoteDefaultCommit())
 	if err != nil {
-		return fmt.Errorf(
-			"the reference %s/%s is not reachable from HEAD: %w",
-			gitcfg.DefaultRemote,
-			gitcfg.DefaultBranch,
-			err,
-		)
+		logger.Debug().Msg(err.Error())
```
Comment (mariux) [inline cmd/terramate/cli/project.go:247]:
```suggestion
		logger.Debug().
		  Msg("A common merge-base can not be determined between HEAD and default branch")
```

File: cmd/terramate/cli/project.go [hunk 12]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
+
 	if mergeBaseCommitID != p.remoteDefaultCommit() {
-		return errors.E(
-			ErrOutdatedLocalRev,
-			"remote %s/%s != HEAD",
-			gitcfg.DefaultRemote,
-			gitcfg.DefaultBranch,
+		logger.Debug().Msgf(
+			"The %s is not the merge-base of current HEAD", remoteDesc,
 		)
```
Comment (mariux) [inline cmd/terramate/cli/project.go:258]: remoteDesc is already in the context of current logger

```suggestion
		logger.Debug().
		  Str("merge_base_hash", mergeBaseCommitID).
		  Msg("The default branch is not equal to the common merge-base of HEAD")
```

File: cmd/terramate/cli/project.go [hunk 13]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
-			err,
-		)
+		logger.Debug().Msg(err.Error())
+		return outOfDateErr
 	}
 
+	logger = logger.With().
+		Str("merge_base", mergeBaseCommitID).
+		Logger()
+
```
Comment (mariux) [inline cmd/terramate/cli/project.go:254]: We do not need to create a logger for a single message

```suggestion
```

File: cmd/terramate/cli/project.go [hunk 14]
```diff
@@ -225,25 +225,38 @@ func (p project) checkDefaultRemote() error {
...
+	logger := log.With().
+		Str("HEAD", p.headCommit()).
+		Str("default_branch", remoteDesc).
+		Str("default_hash", p.remoteDefaultCommit()).
+		Logger()
+
+	outOfDateErr := errors.E(
+		ErrCurrentHeadIsOutOfDate,
+		"Please merge the latest changes from %s into this branch",
+		remoteDesc,
```
Comment (mariux) [inline cmd/terramate/cli/project.go:242]: let's rephrase and also remove inline context.

```suggestion
		"Please update the current branch with the latest changes from the default branch.",
```

**Review by mariux — CHANGES_REQUESTED** [review]

File: cmd/terramate/cli/cli.go [hunk 15]
```diff
@@ -698,6 +682,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
...
 }
 
+func (c *cli) gitSafeguardDefaultBranchIsReachable() {
+	logger := log.With().
+		Str("action", "checkGit()").
+		Logger()
+
+	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
+		return
+	}
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:692]:
```suggestion
	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
		logger.Debug().Msg("Safeguard default-branch-is-reachable is disabled.")
		return
	}
```

File: cmd/terramate/cli/cli.go [hunk 16]
```diff
@@ -698,6 +682,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
 	}
 }
 
+func (c *cli) gitSafeguardDefaultBranchIsReachable() {
+	logger := log.With().
+		Str("action", "checkGit()").
+		Logger()
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:688]:
```suggestion
	logger := log.With().
        Bool("is_repository", c.prj.isRepo).
	    Bool("is_enabled", c.gitSafeguardRemoteEnabled).
		Logger()
```

File: cmd/terramate/cli/cli.go [hunk 17]
```diff
@@ -698,6 +682,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
...
+
+	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
+		return
+	}
+
+	logger.Trace().Msg("check git default remote branch is reachable")
+
+	if err := c.prj.checkRemoteDefaultBranchIsReachable(); err != nil {
+		fatal(err)
+	}
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:698]:
```suggestion
    if err := c.prj.checkRemoteDefaultBranchIsReachable(); err != nil {
	    logger.Trace().Bool("is_reachable", false).Err(err).
	        Msg("Safeguard default-branch-is-reachable failed.")
        fatal(err)
	}
    logger.Trace().Bool("is_reachable", true).
	    Msg("Safeguard default-branch-is-reachable passed.")
```

File: cmd/terramate/cli/cli.go [hunk 18]
```diff
@@ -698,6 +682,22 @@ func (c *cli) gitSafeguards(checks terramate.RepoChecks, shouldAbort bool) {
...
+	logger := log.With().
+		Str("action", "checkGit()").
+		Logger()
+
+	if !c.prj.isRepo || !c.gitSafeguardRemoteEnabled() {
+		return
+	}
+
+	logger.Trace().Msg("check git default remote branch is reachable")
+
```
Comment (mariux) [inline cmd/terramate/cli/cli.go:695]:
```suggestion
```

**Review by mariux — CHANGES_REQUESTED** [review]

File: cmd/terramate/cli/project.go (hunk 9 above)
Comment (i4ki) [reply cmd/terramate/cli/project.go:240]: we use them to match errors in the e2e tests:
```
./cmd/terramate/e2etests/safeguard_test.go:81:			StderrRegex: string(cli.ErrCurrentHeadIsOutOfDate),
./cmd/terramate/e2etests/safeguard_test.go:148:				StderrRegex: string(cli.ErrCurrentHeadIsOutOfDate),
```

**Comment by i4ki** [comment]
@mariux can you have a look again?

**Review by mariux — COMMENTED** [review]
lgtm

**Review by katcipis — APPROVED** [review]

**Pull request merged**